plugins/libplugin-pay.c

```diff
@@ -0,0 +1,780 @@
+{
+	struct out_req *req;
+	struct route_hop *first = &p->route[0];
+	struct secret *secrets;
+	p->createonion_response = tal_createonion_response_from_json(p, buffer, toks);
+
+	req = jsonrpc_request_start(p->plugin, NULL, "sendonion",
+				    payment_sendonion_success,
+				    payment_rpc_failure, p);
+	json_add_hex_talarr(req->js, "onion", p->createonion_response->onion);
+
+	json_object_start(req->js, "first_hop");
+	json_add_short_channel_id(req->js, "channel", &first->channel_id);
+	json_add_num(req->js, "direction", first->direction);
+	json_add_amount_msat_only(req->js, "amount_msat", first->amount);
+	json_add_num(req->js, "delay", first->delay);
+	json_add_node_id(req->js, "id", &first->nodeid);
+	json_object_end(req->js);
+
+	json_add_sha256(req->js, "payment_hash", p->payment_hash);
+
+	json_array_start(req->js, "shared_secrets");
+	secrets = p->createonion_response->shared_secrets;
+	for(size_t i=0; i<tal_count(secrets); i++)
+		json_add_secret(req->js, NULL, &secrets[i]);
+	json_array_end(req->js);
+
+	json_add_num(req->js, "partid", p->partid);
+
+	send_outreq(p->plugin, req);
+	return command_still_pending(cmd);
+}
+
+/* Temporary serialization method for the tlv_payload.data until we rework the
+ * API that is generated from the specs to use the setter/getter interface. */
+static void tlvstream_set_tlv_payload_data(struct tlv_field **stream,
+					   struct secret *payment_secret,
+					   u64 total_msat)
+{
+		u8 *ser = tal_arr(NULL, u8, 0);
+		towire_secret(&ser, payment_secret);
+		towire_tu64(&ser, total_msat);
+		tlvstream_set_raw(stream, 8, take(ser));
+}
+
+static void payment_compute_onion_payloads(struct payment *p)
+{
+	struct createonion_request *cr;
+	size_t hopcount;
+	static struct short_channel_id all_zero_scid;
+	struct createonion_hop *cur;
+	struct payment *root = payment_root(p);
+	p->step = PAYMENT_STEP_ONION_PAYLOAD;
+	hopcount = tal_count(p->route);
+
+	/* Now compute the payload we're about to pass to `createonion` */
+	cr = p->createonion_request = tal(p, struct createonion_request);
+	cr->assocdata = tal_arr(cr, u8, 0);
+	towire_sha256(&cr->assocdata, p->payment_hash);
+	cr->session_key = NULL;
+	cr->hops = tal_arr(cr, struct createonion_hop, tal_count(p->route));
+
+	/* Non-final hops */
+	for (int i = 0; i < hopcount - 1; i++) {
+		/* The message is destined for hop i, but contains fields for
+		 * i+1 */
+		cur = &cr->hops[i];
+		cur->style = p->route[i].style;
+		cur->pubkey = p->route[i].nodeid;
+		switch (cur->style) {
+		case ROUTE_HOP_LEGACY:
+			cur->legacy_payload =
+			    tal(cr->hops, struct legacy_payload);
+			cur->legacy_payload->forward_amt =
+			    p->route[i + 1].amount;
+			cur->legacy_payload->scid = p->route[i + 1].channel_id;
+			cur->legacy_payload->outgoing_cltv =
+			    p->start_block + p->route[i + 1].delay;
+			break;
+		case ROUTE_HOP_TLV:
+			cur->tlv_payload = tlv_tlv_payload_new(cr->hops);
+			tlvstream_set_tu64(
+			    &cur->tlv_payload->fields, 2,
```

Comment:
are there conts/enums we can use for the type number instead of the straight int?

Reply:
I added the code to generate an `enum` in the `gen_onion_wire.h` file, so we can use the names. Thanks for the suggestion ^^